ui: Type-safe Local Settings in redux state

mrtracy opened the pull request:
Adds LocalSetting, a type-safe structure for accessing "Local Settings" in the
Redux state. Previously, local settings were accessed and set directly in a way
that had no type safety; now, setting or accessing a single local setting is
type safe (the setting for each individual setting can be different).

Also adding some clarifying comments to explain the use of local settings -
essentially, there are some simple settings which should be stored in the redux
state (so that they are kept for the session), but are only ever accessed by a
single component, and the settings dictionary is a convenient way to do this
without requiring the creation of new redux boilerplate.

Renames redux/ui.tx -> redux/localsettings.ts, a more descriptive name which is
less ambiguous from uiData.ts.

This was being developed as part of a larger PR around alerts, but was getting
large and required refactors of unrelated components, so I have broken it out
into its own commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14588)
<!-- Reviewable:end -->

Commit:
ui: Type-safe Local Settings in redux state

Adds LocalSetting, a type-safe structure for accessing "Local Settings" in the
Redux state. Previously, local settings were accessed and set directly in a way
that had no type safety; now, setting or accessing a single local setting is
type safe (the setting for each individual setting can be different).

Also adding some clarifying comments to explain the use of local settings -
essentially, there are some simple settings which should be stored in the redux
state (so that they are kept for the session), but are only ever accessed by a
single component, and the settings dictionary is a convenient way to do this
without requiring the creation of new redux boilerplate.

Renames redux/ui.tx -> redux/localsettings.ts, a more descriptive name which is
less ambiguous from uiData.ts.

This was being developed as part of a larger PR around alerts, but was getting
large and required refactors of unrelated components, so I have broken it out
into its own commit.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/containers/events.tsx, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7-OeCIy4X7Aq4uWe:-Kgq7-OeCIy4X7Aq4uWf:bhgeiyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/events.tsx#L23)):*
> ```XML
> const EVENT_BOX_NUM_EVENTS = 10;
> 
> // LocalSetting which maintains table sort order.
> ```

remove all comments of this form, they aren't useful because they restate the (very simple) code that follows.

---

*[pkg/ui/src/containers/nodesOverview.tsx, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7BBFadXKJwU7CWGq:-Kgq7BBFadXKJwU7CWGr:b32ww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/nodesOverview.tsx#L21)):*
> ```XML
> import { NodeStatus$Properties, MetricConstants, BytesUsed } from  "../util/proto";
> 
> // LocalSetting which maintains live table sort order.
> ```

rm

---

*[pkg/ui/src/containers/nodesOverview.tsx, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7CZK_ANP35vFb-ks:-Kgq7CZK_ANP35vFb-kt:b32ww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/nodesOverview.tsx#L25)):*
> ```XML
>   "nodes/live_sort_setting", (s) => s.ui,
> );
> // LocalSetting which maintains dead table sort order.
> ```

rm

---

*[pkg/ui/src/containers/timescale.tsx, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7FB9GbVd_BO1oeGK:-Kgq7FB9GbVd_BO1oeGL:b-tirg9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/timescale.tsx#L16)):*
> ```XML
> import { LongToMoment } from "../util/convert";
> 
> // Tracks whether the default timescale been set once in the app. Tracked across
> ```

This one is useful, unlike the others!

---

*[pkg/ui/src/containers/databases/databaseGrants.tsx, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7J9ZErZGViAXOiCK:-Kgq7J9ZErZGViAXOiCL:b32ww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/databaseGrants.tsx#L28)):*
> ```XML
> export const DatabaseGrantsSortedTable = SortedTable as new () => SortedTable<protos.cockroach.server.serverpb.DatabaseDetailsResponse.Grant>;
> 
> // LocalSetting which maintains table sort order.
> ```

rm

---

*[pkg/ui/src/containers/databases/databaseTables.tsx, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6y4mpg9oInI3d_kk:-Kgq6y4mpg9oInI3d_kl:b32ww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/databaseTables.tsx#L24)):*
> ```XML
> } from "./databaseSummary";
> 
> // LocalSetting which maintains table sort order.
> ```

rm

---

*[pkg/ui/src/containers/databases/tableDetails.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6uUU0J3Y6FUk_jWC:-Kgq6uUU0J3Y6FUk_jWD:b32ww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/tableDetails.tsx#L26)):*
> ```XML
> export const GrantsSortedTable = SortedTable as new () => SortedTable<protos.cockroach.server.serverpb.TableDetailsResponse.Grant$Properties>;
> 
> // LocalSetting which maintains table sort order.
> ```

rm

---

*[pkg/ui/src/redux/localsettings.spec.ts, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6N2leqewSm_KXStS:-Kgq6N2leqewSm_KXStT:b-7soxve) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.spec.ts#L22)):*
> ```XML
>   });
> 
>   describe("reducer", function() {
> ```

"ui reducer reducer"?

---

*[pkg/ui/src/redux/localsettings.spec.ts, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6V8AU00ty875Eb80:-Kgq6V8AU00ty875Eb81:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.spec.ts#L124)):*
> ```XML
>       assert.equal(numberSetting.selector(topLevelState), 5);
>     });
> ```

remove

---

*[pkg/ui/src/redux/localsettings.ts, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq5r-eZvmZxsN-gyzK:-Kgq5r-fRO_CQn3Yied0:b-3k82nz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.ts#L19)):*
> ```XML
> const SET_UI_VALUE = "cockroachui/ui/SET_UI_VALUE";
> 
> export class LocalSettingData {
> ```

why classes rather than type aliases?

---

*[pkg/ui/src/redux/localsettings.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq5vfVPmg-kU_ydcrW:-Kgq5vfVPmg-kU_ydcrX:b-lh5vaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.ts#L35)):*
> ```XML
>  */
> export default function reducer(state: LocalSettingsDict, action: Action): LocalSettingsDict {
>   if (action === undefined) {
> ```

are these necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-:-Kgq7O8hCiQPMZpoYuW9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 7 of 15 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/ui/src/containers/events.tsx, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7-OeCIy4X7Aq4uWe:-Kgq8FKCZDu0L_K3rMgu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/events.tsx#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove all comments of this form, they aren't useful because they restate the (very simple) code that follows.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/nodesOverview.tsx, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7BBFadXKJwU7CWGq:-Kgq8Fexwx1MfPWqdmX4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/nodesOverview.tsx#L21)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

rm
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/nodesOverview.tsx, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7CZK_ANP35vFb-ks:-Kgq8G1_HSELsPlbJwW1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/nodesOverview.tsx#L25)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

rm
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/databases/databaseGrants.tsx, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq7J9ZErZGViAXOiCK:-Kgq8HXinoT1CL9I6R6H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/databaseGrants.tsx#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

rm
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/databases/databaseTables.tsx, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6y4mpg9oInI3d_kk:-Kgq8HuT8DtbTNA4VD2h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/databaseTables.tsx#L24)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

rm
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/databases/tableDetails.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6uUU0J3Y6FUk_jWC:-Kgq8IOMkkvMZDrP-fkM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/containers/databases/tableDetails.tsx#L26)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

rm
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/localsettings.spec.ts, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6N2leqewSm_KXStS:-Kgq8J2edS0VLWcpM-28:b-aebb2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.spec.ts#L22)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"ui reducer reducer"?
</blockquote></details>

Fixed these labels up a bit.

---

*[pkg/ui/src/redux/localsettings.spec.ts, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq6V8AU00ty875Eb80:-Kgq8O5fEpJoPTd_C1l3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.spec.ts#L124)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/localsettings.ts, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq5r-eZvmZxsN-gyzK:-Kgq8akSbc4lps948m8v:bc4r3mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.ts#L19)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why classes rather than type aliases?
</blockquote></details>

Didn't notice they were classes, changed to interfaces; I typically favor interfaces than type aliases, I believe it to be more idiomatic.

---

*[pkg/ui/src/redux/localsettings.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq5vfVPmg-kU_ydcrW:-Kgq9LUpvCYsZAjEIxFd:bny505x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.ts#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are these necessary?
</blockquote></details>

Redux reducers define their own shape, and that includes returning a default non-null zero value if the provided state is undefined.

I've changed this up to both return something if "Action" is nil, and I am now using the `_.isNil()` lodash method which is appropriate for flexibly detecting undefined values on objects. God bless javascript and its complicated three-valued states.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/ui/src/redux/localsettings.ts, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588#-Kgq5r-eZvmZxsN-gyzK:-KgqD0MwLEOwlbFmvZOk:bfcc4gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8c89f6baadd8ae01bd541f831bc861749ccdfd/pkg/ui/src/redux/localsettings.ts#L19)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Didn't notice they were classes, changed to interfaces; I typically favor interfaces than type aliases, I believe it to be more idiomatic.
</blockquote></details>

I know that has been the typical approach, but I wonder if if that should be rethought; I haven't seen any indication that interfaces are more idiomatic. Not blocking for this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14588)*
<!-- Sent from Reviewable.io -->
